### ISSUE TITLE
Is Conda Cache thread-safe?

### ISSUE THREAD

**Issue by cc636489** [conda/conda#14458]
### Checklist

- [x] I added a descriptive title
- [x] I searched open requests and couldn't find a duplicate

Duplicate of: #13055

### What is the idea?

Hi everyone,

I’ve recently started using Conda for managing Python environments and packages. While exploring its features, I couldn't find detailed information about its cache mechanism. Specifically, I’d like to know:

- Is the Conda cache thread-safe?
- If it is, does it implement any locking mechanism to prevent simultaneous writes or reads from corrupting the cache?

Understanding this is crucial for using Conda in multi-threaded or concurrent processes. Any insights, documentation links, or user experiences would be greatly appreciated!

Thanks in advance!

### Why is this needed?

We prefer to re-use the cache in CI environment.

### What should happen?

It would be great if Conda Cache is thread-safe, for speeding up the build process of CI.

**Comment by travishathaway** [comment]
@cc636489,

I think what you meant to ask is whether conda can be run in parallel (i.e. with multiple processes on the same machine). The answer to that question is no it cannot. Please see this following issues for discussion on this topic:

- https://github.com/conda/conda/issues/13055
- https://github.com/conda/conda/issues/13037

If you need more specific information about pixi, please head over to their issue queue:

- https://github.com/prefix-dev/pixi/issues

**Comment by cc636489** [comment]
I modified the content and yes, I meant to ask whether Conda can be run in parallel. Thanks for your answer! The references you provided did help my understanding! Much appreciated!

**Comment by wolfv** [comment]
@cc636489 in `rattler` we implemented a multi-process and thread-safe cache: https://github.com/conda/rattler

You can use `py-rattler` Python bindings if you want to create environments programatically.

**Comment by cc636489** [comment]
Hi @wolfv ,

Happy new year! And thanks a lot for your reply and providing an alternative approach to work with Conda ecosystem! As far as I could understand from the documentation of rattler, I have two follow-up questions:

- Since rattler is a library which can speed-up the overall process to work with conda ecosystem (i.e. it programmatically provides an optimized solver and installer, etc.), I am wondering, besides it provides APIs for Python users, any gifts for C++ users?

- How does rattler handle its cache? Does rattler directly work with conda cache, i.e. while installing a package via rattler, it creates conda/conda-compatiable cache and ensures its thread-safe access?  

Thanks in advance for your time!

**Comment by wolfv** [comment]
Hi @cc636489 - we have Python bindings, but no C++ bindings yet. Another of our projects (resolvo) does have C++ bindings. If you are willing, we would be happy to help you to create C++ bindings for rattler.

We have our own cache that works a bit different from conda's cache (we use special files to lock the directory for read / write access).
